### PR TITLE
Fix question subdirs to work with web editor

### LINKED ISSUE #2593: Question subdirs do not fully work with web editor
Following on from #2489, questions in subdirectories have the following bugs:

* QID renaming on the "Settings" tab currently fails with an error to use only approved characters.
* There is no way to move questions between subdirectories or to create new subdirectories (this is related to the point above).
* When copying a question from another course, the subdirectory information is dropped.

One way to fix the first two issues would be to allow renaming to any path like `subdir/childdir/qname`, which would create `subdir/childdir` if needed and move `qname` to that directory.

### PR DESCRIPTION
**Pull request by nicknytko**
Fixes #2593 

QID, AID, CIID can now be renamed to have forward slashes in them on the web editor, indicating subdirectories.  Renaming questions that are already nested will collapse empty subfolders.
Copying questions from other courses will retain folder nesting structure.
Deleting questions/assessments/course instances that are nested will remove any empty parent directories to prevent sync issues.

### CODE REVIEW

**Comment by nicknytko** [comment]
Tagging @tdy and @mfsilva22 for review, would you two mind running this and trying to break the web editor?  Things like renaming questions/assessments to have subdirectories, deleting nested questions, etc.

**Comment by tdy** [comment]
FYI, in case you think it's worth documenting, web editor renaming doesn't play well with nodemon.

**Comment by nicknytko** [comment]
> FYI, in case you think it's worth documenting, web editor renaming doesn't play well with nodemon.

I thought questions, assessments, etc. should be okay because those force a course sync on edit?  Or do you mean things like course elements which I think don't update right away on local instances?

**Comment by nicknytko** [comment]
@tdy I've factored out the nodemon stuff into a config section in `package.json`: (https://github.com/PrairieLearn/PrairieLearn/pull/2598/commits/df51db94ed4ab88c482e8662d41c38995fcd67dc)

**Comment by tdy** [comment]
I mean something like this:

1. `/course/3/question/86/settings`
2. Click `Change QID` and change `addNumbers` -> `alpha/bravo/charlie/addNumbers`
3. Click `Change`
4. The page tries to reload at this point (or something?), but nodemon has already detected filesystem changes and has started rebooting node, so you just get a 404

> 
> 
> @tdy I've factored out the nodemon stuff into a config section in `package.json`: ([df51db9](https://github.com/PrairieLearn/PrairieLearn/commit/df51db94ed4ab88c482e8662d41c38995fcd67dc))

Nice! Works on my end.

**Comment by tdy** [comment]
Oops, my initial reply had a leftover draft that I'd started yesterday.

**Comment by tdy** [comment]
This worked for me last week, but now I'm getting this error (renaming `addNumbers` -> `a/b/addNumbers`:

<details>
<summary>console output</summary>
error: Error in write() EACCES: permission denied, rename '/PrairieLearn/testCourse/questions/addNumbers' -> '/PrairieLearn/testCourse/questions/a/b/addNumbers' {"errno":-13,"code":"EACCES","syscall":"rename","path":"/PrairieLearn/testCourse/questions/addNumbers","dest":"/PrairieLearn/testCourse/questions/a/b/addNumbers","stack":"Async Stacktrace:\n    at /PrairieLearn/lib/editors.js:171:29\n    at /PrairieLearn/lib/editors.js:1173:17\n    at /PrairieLearn/lib/editors.js:1101:25\n\nError: EACCES: permission denied, rename '/PrairieLearn/testCourse/questions/addNumbers' -> '/PrairieLearn/testCourse/questions/a/b/addNumbers'"}
error: Error in doEdit() edit failed {"stack":"Async Stacktrace:\n    at /PrairieLearn/pages/instructorQuestionSettings/instructorQuestionSettings.js:60:25\n\nError: edit failed\n    at Object._finishWithFailure (/PrairieLearn/lib/editors.js:385:26)\n    at /PrairieLearn/lib/server-jobs.js:129:34\n    at /PrairieLearn/node_modules/@prairielearn/prairielib/lib/sql-db.js:458:9\n    at Query.<anonymous> (/PrairieLearn/node_modules/@prairielearn/prairielib/lib/sql-db.js:429:17)\n    at Query.handleReadyForQuery (/PrairieLearn/node_modules/pg/lib/query.js:129:12)\n    at Connection.<anonymous> (/PrairieLearn/node_modules/pg/lib/client.js:266:19)\n    at Connection.emit (events.js:315:20)\n    at Connection.EventEmitter.emit (domain.js:482:12)\n    at /PrairieLearn/node_modules/pg/lib/connection.js:109:10\n    at Parser.parse (/PrairieLearn/node_modules/pg-protocol/dist/parser.js:42:17)\n    at Socket.<anonymous> (/PrairieLearn/node_modules/pg-protocol/dist/index.js:8:42)\n    at Socket.emit (events.js:315:20)\n    at Socket.EventEmitter.emit (domain.js:482:12)\n    at addChunk (_stream_readable.js:295:12)\n    at readableAddChunk (_stream_readable.js:271:9)\n    at Socket.Readable.push (_stream_readable.js:212:10)"}
</details>

File: lib/editors.js [hunk 1]
```diff
@@ -389,6 +389,97 @@ class Editor {
...
+                    } else {
+                        debug('No files, deleting', delPath);
+                        fs.remove(delPath, (err) => {
+                            if (ERR(err, callback)) return;
+                            callback(null);
+                        });
+                    }
+                });
+            } else { callback(null); }
+        }, callback);
```
Comment (mwest1066) [inline lib/editors.js:430]: See the second bullet point here: https://prairielearn.readthedocs.io/en/latest/dev-guide/#stack-traces-with-callback-style-functions

**Pull request merged**